internal/providers/amazon/eks/workflow/create_asg_activity.go

```diff
@@ -0,0 +1,304 @@
+		},
+		{
+			ParameterKey:   aws.String("NodeAutoScalingGroupMinSize"),
+			ParameterValue: aws.String(fmt.Sprintf("%d", input.NodeMinCount)),
+		},
+		{
+			ParameterKey:   aws.String("NodeAutoScalingGroupMaxSize"),
+			ParameterValue: aws.String(fmt.Sprintf("%d", input.NodeMaxCount)),
+		},
+		{
+			ParameterKey:   aws.String("NodeAutoScalingInitSize"),
+			ParameterValue: aws.String(fmt.Sprintf("%d", input.Count)),
+		},
+		{
+			ParameterKey:   aws.String("ClusterName"),
+			ParameterValue: aws.String(input.ClusterName),
+		},
+		{
+			ParameterKey:   aws.String("NodeGroupName"),
+			ParameterValue: aws.String(input.Name),
+		},
+		{
+			ParameterKey:   aws.String("ClusterControlPlaneSecurityGroup"),
+			ParameterValue: aws.String(input.SecurityGroupID),
+		},
+		{
+			ParameterKey:   aws.String("NodeSecurityGroup"),
+			ParameterValue: aws.String(input.NodeSecurityGroupID),
+		},
+		{
+			ParameterKey:   aws.String("VpcId"),
+			ParameterValue: aws.String(input.VpcID),
+		}, {
+			ParameterKey:   aws.String("Subnets"),
+			ParameterValue: aws.String(strings.Join(subnetIDs, ",")),
+		},
+		{
+			ParameterKey:   aws.String("NodeInstanceRoleId"),
+			ParameterValue: aws.String(input.NodeInstanceRoleID),
+		},
+		{
+			ParameterKey:   aws.String("ClusterAutoscalerEnabled"),
+			ParameterValue: aws.String(fmt.Sprint(clusterAutoscalerEnabled)),
+		},
+		{
+			ParameterKey:   aws.String("TerminationDetachEnabled"),
+			ParameterValue: aws.String(fmt.Sprint(terminationDetachEnabled)),
+		},
+		{
+			ParameterKey:   aws.String("BootstrapArguments"),
+			ParameterValue: aws.String(fmt.Sprintf("--kubelet-extra-args '--node-labels %v'", strings.Join(nodeLabels, ","))),
+		},
+	}
+	clientRequestToken := generateRequestToken(input.AWSClientRequestTokenBase, CreateAsgActivityName)
+
+	createStackInput := &cloudformation.CreateStackInput{
+		ClientRequestToken: aws.String(clientRequestToken),
+		DisableRollback:    aws.Bool(true),
+		StackName:          aws.String(input.StackName),
+		Capabilities:       []*string{aws.String(cloudformation.CapabilityCapabilityIam)},
+		Parameters:         stackParams,
+		Tags:               tags,
+		TemplateBody:       aws.String(input.CloudFormationTemplate),
+		TimeoutInMinutes:   aws.Int64(10),
+	}
+	_, err = cloudformationClient.CreateStack(createStackInput)
+	if err != nil {
+		return nil, errors.WrapIff(err, "could not create '%s' CF stack", input.StackName)
+	}
+
+	describeStacksInput := &cloudformation.DescribeStacksInput{StackName: aws.String(input.StackName)}
+
+	err = errors.WrapIff(cloudformationClient.WaitUntilStackCreateComplete(describeStacksInput),
+		"waiting for %q CF stack create operation to complete failed", input.StackName)
+	err = pkgCloudformation.NewAwsStackFailure(err, input.StackName, clientRequestToken, cloudformationClient)
+	if err != nil {
+		return nil, err
+	}
+
+	// wait for ASG fulfillment
+	err = a.waitForASGToBeFulfilled(ctx, logger, awsSession, input.StackName, input.Name)
```

Comment:
What would that buy us?

Reply:
High activity timeout (even with retries) causes an activity to be blocked in case of a worker restart until the timeout is reached. With a heartbeat timeout the activity can be timed out way earlier, while the original timeout of the activity remains intact.

Reply:
Need to think this through as if this is in a separate activity which polls the ASG every 5 seconds then it will create a new AWS session every 5 seconds. Amazon's recommendation is to create the session and re-use it across multiple client calls for performance reason.